src/native/libs/System.IO.Compression.Native/zlib_allocator_win.c

```diff
@@ -0,0 +1,93 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <Windows.h>
+#include <heapapi.h>
+#include <intsafe.h>
+#include <crtdbg.h> /* _ASSERTE */
+
+#include <zlib_allocator.h>
+
+/* A custom allocator for zlib that uses a dedicated heap.
+   This provides better performance and avoids fragmentation
+   that can occur on Windows when using the default process heap.
+ */
+
+// Gets the special heap we'll allocate from.
+HANDLE GetZlibHeap()
+{
+#ifdef _WIN64
+    static HANDLE s_hPublishedHeap = NULL;
+
+    // If already initialized, return immediately.
+    // We don't need a volatile read here since the publish is performed with release semantics.
+    if (s_hPublishedHeap != NULL) { return s_hPublishedHeap; }
+
+    // Attempt to create a new heap. The heap will be dynamically sized.
+    HANDLE hNewHeap = HeapCreate(0, 0, 0);
+
+    if (hNewHeap != NULL)
+    {
+        // We created a new heap. Attempt to publish it.
+        if (InterlockedCompareExchangePointer(&s_hPublishedHeap, hNewHeap, NULL) != NULL)
+        {
+            HeapDestroy(hNewHeap); // Somebody published before us. Destroy our heap.
+            hNewHeap = NULL; // Guard against accidental use later in the method.
+        }
+    }
+    else
+    {
+        // If we can't create a new heap, fall back to the process default heap.
+        InterlockedCompareExchangePointer(&s_hPublishedHeap, GetProcessHeap(), NULL);
+    }
+
+    // Some thread - perhaps us, perhaps somebody else - published the heap. Return it.
+    // We don't need a volatile read here since the publish is performed with release semantics.
+    _ASSERTE(s_hPublishedHeap != NULL);
+    return s_hPublishedHeap;
+#else
+    // We don't want to create a new heap in a 32-bit process because it could end up
+    // reserving too much of the address space. Instead, fall back to the normal process heap.
+    return GetProcessHeap();
+#endif
+}
+
+voidpf z_custom_calloc(opaque, items, size)
+    voidpf opaque;
+    unsigned items;
+    unsigned size;
+{
+    (void)opaque; // suppress C4100 - unreferenced formal parameter
+
+    // If initializing a fixed-size structure, zero the memory.
+    DWORD dwFlags = (items == 1) ? HEAP_ZERO_MEMORY : 0;
+
+    SIZE_T cbRequested;
+    if (sizeof(items) + sizeof(size) <= sizeof(cbRequested))
+    {
+        // multiplication can't overflow; no need for safeint
+        cbRequested = (SIZE_T)items * (SIZE_T)size;
+    }
+    else
+    {
+        // multiplication can overflow; go through safeint
+        if (FAILED(SIZETMult(items, size, &cbRequested))) { return NULL; }
+    }
+
+    return HeapAlloc(GetZlibHeap(), dwFlags, cbRequested);
```

Comment:
@jkotas what is your thinking on alignment of memory returned by HeapAlloc?  

I see zlib-ng expects 16 or 64 byte aligned memory.  https://github.com/dotnet/runtime/blob/7721695e214fc3c58c280197fa9127e792b765a3/src/native/external/zlib-ng/zutil_p.h#L17-L35

HealAlloc will return 16-byte aligned memory on x64, but 8-byte aligned memory on x86.